pkg/providers/tinkerbell/controlplane.go

```diff
@@ -0,0 +1,204 @@
+package tinkerbell
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	tinkerbellv1 "github.com/tinkerbell/cluster-api-provider-tinkerbell/api/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/clients/kubernetes"
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/clusterapi"
+	yamlcapi "github.com/aws/eks-anywhere/pkg/clusterapi/yaml"
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/providers"
+	"github.com/aws/eks-anywhere/pkg/providers/tinkerbell/hardware"
+	"github.com/aws/eks-anywhere/pkg/yamlutil"
+)
+
+// BaseControlPlane represents a CAPI Tinkerbell control plane.
+type BaseControlPlane = clusterapi.ControlPlane[*tinkerbellv1.TinkerbellCluster, *tinkerbellv1.TinkerbellMachineTemplate]
+
+// ControlPlane holds the Tinkerbell specific objects for a CAPI Tinkerbell control plane.
+type ControlPlane struct {
+	BaseControlPlane
+	Secrets []*corev1.Secret
+}
+
+// Objects returns the control plane objects associated with the Tinkerbell cluster.
+func (p ControlPlane) Objects() []kubernetes.Object {
+	o := p.BaseControlPlane.Objects()
+	o = getSecrets(o, p.Secrets)
+
+	return o
+}
+
+func getSecrets(o []kubernetes.Object, secrets []*corev1.Secret) []kubernetes.Object {
+	for _, s := range secrets {
+		o = append(o, s)
+	}
+	return o
+}
+
+func getControlPlaneMachineSpec(clusterSpec *cluster.Spec, diskExtractor *hardware.DiskExtractor) (*anywherev1.TinkerbellMachineConfigSpec, error) {
+	var controlPlaneMachineSpec *anywherev1.TinkerbellMachineConfigSpec
+	if clusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef != nil && clusterSpec.TinkerbellMachineConfigs[clusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name] != nil {
+		controlPlaneMachineSpec = &clusterSpec.TinkerbellMachineConfigs[clusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec
+		err := diskExtractor.Register(controlPlaneMachineSpec.HardwareSelector)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return controlPlaneMachineSpec, nil
+}
+
+func getWorkerNodeGroupMachineSpec(clusterSpec *cluster.Spec, diskExtractor *hardware.DiskExtractor) (map[string]anywherev1.TinkerbellMachineConfigSpec, error) {
```

Comment:
Should we move this to a `workers.go` file since it should be separate from cp objects? Also double-check with @tatlat to ensure that this doesn't conflict with his workers task

Reply:
So I actually already have something like this for my task. I don't think templateBulider.GenerateCAPISpecControlPlane() uses WorkerNodeGroupMachineSpecs. I might be wrong but I think you could probably pass a nil value and it would still generate the control plane objects. Maybe it would be better to put all the methods for getting machine specs (cp, etcd, workers) and generating the template builder into one file that we could share.

Reply:
sounds good. I'll move the template builder out to a common methods